pkg/backends/calico/routes.go

```diff
@@ -15,30 +15,240 @@ package calico
+// necessary for monitoring the services/endpoints resources for
+// valid service ips to advertise
 type routeGenerator struct {
-	client *client
+	sync.Mutex
+	client                  *client
+	nodeName                string
+	svcInformer, epInformer cache.Controller
+	svcIndexer, epIndexer   cache.Indexer
+	svcRouteMap             map[string]string // maps service name to ip
 }
 
+// NewRouteGenerator initializes a kube-api client and the informers
 func NewRouteGenerator(c *client) (rg *routeGenerator, err error) {
-	rg = &routeGenerator{client: c}
+	log.Debug("+NewRouteGenerator")
+	// initialize empty route generator
+	rg = &routeGenerator{
+		client:      c,
+		nodeName:    template.NodeName,
+		svcRouteMap: make(map[string]string),
+	}
+
+	// set up k8s client
+	// attempt 1: KUBECONFIG env var
+	cfgFile := os.Getenv("KUBECONFIG")
+	cfg, err := clientcmd.BuildConfigFromFlags("", cfgFile)
+	if err != nil {
+		// attempt 2: in cluster config
+		if cfg, err = rest.InClusterConfig(); err != nil {
+			return nil, err
+		}
+	}
+	client, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// set up services informers
+	svcWatcher := cache.NewListWatchFromClient(client.Core().RESTClient(), "services", "", fields.Everything())
+	svcHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onSvcAdd, UpdateFunc: rg.onSvcUpdate, DeleteFunc: rg.onSvcDelete}
+	rg.svcIndexer, rg.svcInformer = cache.NewIndexerInformer(svcWatcher, &v1.Service{}, 0, svcHandler, cache.Indexers{})
+
+	// set up endpoints informers
+	epWatcher := cache.NewListWatchFromClient(client.Core().RESTClient(), "endpoints", "", fields.Everything())
+	epHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onEPAdd, UpdateFunc: rg.onEPUpdate, DeleteFunc: rg.onEPDelete}
+	rg.epIndexer, rg.epInformer = cache.NewIndexerInformer(epWatcher, &v1.Endpoints{}, 0, epHandler, cache.Indexers{})
+
+	log.Debug("-NewRouteGenerator")
 	return
 }
 
 func (rg *routeGenerator) Start() (err error) {
 	// MVP implementation: read CIDRs to advertise,
 	// comma-separated, from an environment variable
 	// CALICO_STATIC_ROUTES.
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
-	cidrs := []string{}
-	for _, route := range strings.Split(routeString, ",") {
-		cidr := strings.TrimSpace(route)
-		if cidr != "" {
-			cidrs = append(cidrs, cidr)
+	ch := make(chan struct{})
+	go rg.svcInformer.Run(ch)
+	go rg.epInformer.Run(ch)
+
+	/*
+		routeString := os.Getenv("CALICO_STATIC_ROUTES")
+		cidrs := []string{}
+		for _, route := range strings.Split(routeString, ",") {
+			cidr := strings.TrimSpace(route)
+			if cidr != "" {
+				cidrs = append(cidrs, cidr)
+			}
+		}
+		rg.client.updateRoutes(cidrs)
+	*/
+	return
+}
+
+// setRouteForSvc handles the main logic to check if a specified service or endpoint
+// should have its route advertised by the node running this code
+func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
+	log.Debug("+setRouteForSvc")
+	rg.Lock()
```

Comment:
Do we need to lock for the entire operation? It seems this lock is just for simultaneous writes to the routemap - could we instead only hold the lock while we're accessing that map? I think that's, in general, a nicer approach and makes it more clear which state is modified by multiple routines.

We could pull out the modifications into helper functions, which would ensure we're always locking when making changes.

e.g.
```go
func advertiseServiceRoutes(svc *v1.Service) {}
func revokeServiceRoutes(name string) {}
```
And those funcs would do the right locking.